cucumber-messages/messages.proto

```diff
@@ -428,6 +428,9 @@ message TestCaseStarted {
   string pickleId = 1;
   Timestamp timestamp = 2;
   Platform platform = 3;
+  // Used if the implementation has implemented retry
+  //   also present in TestStepStarted, TestStepFinished, TestCaseFinished
+  uint32 attemptNumber = 4;
```

Comment:
Is this index-0 or index-1 based? The `Number` suffix creates ambiguity here. So I think this should just be named `attempt`.

I don't quite like that this creates a composite key either in all test case events. It makes dealing with these events harder. Perhaps we can replace the `pickleId` in `TestStepStarted`, `TestStepFinished`, `TestCaseFinished` with a  `testCaseId`. Then we would only need to include the `pickleId` and `attempt` in `TestCaseStarted`.

Reply:
This is 1 based. How does the presence of number create more / less ambiguity?

So to implement retry, we currently only fire a single TestCasePrepared. Then for each attempt we trigger TestCaseStarted / test step events / TestCaseFinished. Thus we need the composite key of pickleId and attemptNumber for any attempt at running a test case.

Reply:
I think I prefer the composite key because it can be used by languages that have implemented retry and for those that don't they have anything more complex

Reply:
> This is 1 based. How does the presence of number create more / less ambiguity?

I assumed it would be a 0 based index, it is conventional and makes it easier to work with - no need to subtract `-1` to convert it to a list index.

The other thing with `Number` is that it is stuttering, the fact that it is a number is already included in the definition `uint32 attemptNumber = 4;`. Adding `Number` makes it stutter a bit. 

> I think I prefer the composite key because it can be used by languages that have implemented retry and for those that don't they have anything more complex

We should design interfaces for the benefit of the consumer. With a unique identifier naive consumers can display a stream of test cases and their result. More advanced consumers can then collate test cases by feature, by retry, or some other criteria.

Reply:
I think we should triangulate here - generate some pull from another angle. We could try to render this information in [cucumber-react](https://github.com/cucumber/cucumber/tree/master/cucumber-react/javascript). If that turns out to be easy we already have a good model. If it feels wrong, we can propose an improvement.

We've done this in the past and it seems to work reasonably well.

Reply:
I'm fine making it 0 based and removing the "Number" suffix.

Given the current situation I don't know if adding another key will make it more or less complex.

And okay you convinced me. Moving to a testCaseId.